### PR TITLE
Reduce latency of fuzz target recording for testcase upload

### PR DESCRIPTION
**Pull request by hogo6002**
Adding a trusted uploader allowlist to bypass the fuzz target check.
Resolve https://github.com/google/clusterfuzz/issues/3005

### CODE REVIEW

File: src/appengine/handlers/upload_testcase.py [hunk 1]
```diff
@@ -183,7 +188,10 @@ def find_fuzz_target(engine, target_name, job_name):
 
   target = data_handler.get_fuzz_target(candidate_name)
   if not target:
-    raise helpers.EarlyExitException('Fuzz target does not exist.', 400)
+    if email and _is_trusted_uploader_allowed(email):
```
Comment (oliverchang) [inline src/appengine/handlers/upload_testcase.py:191]: Can you move this check out of this function? It's a side effect that doesn't fit well with this `find_fuzz_target` function. 

Instead, we could move this to line 355 right?

**Review by oliverchang — COMMENTED** [review]
nice!

**Comment by oliverchang** [comment]
Also, please add a test to https://github.com/google/clusterfuzz/blob/3ad38f4d07e06aef4aa015f6dfe5e9112d8ad5c8/src/clusterfuzz/_internal/tests/appengine/handlers/upload_testcase_test.py.

File: src/appengine/handlers/upload_testcase.py [hunk 2]
```diff
@@ -56,6 +56,11 @@ def _is_uploader_allowed(email):
   return external_users.is_upload_allowed_for_user(email)
 
 
+def _is_trusted_uploader_allowed(email):
```
[the remark this answers is not in the document]
Comment (Alan32Liu) [reply src/appengine/handlers/upload_testcase.py:59]: +1, here is [a docstring guide](https://g3doc.corp.google.com/eng/doc/devguide/py/style/index.md?cl=head#function-docs).

File: src/clusterfuzz/_internal/tests/appengine/handlers/upload_testcase_test.py [hunk 3]
```diff
@@ -487,3 +487,29 @@ def test_external_duplicate(self):
         'timestamp': datetime.datetime(2021, 1, 1, 0, 0),
         'uploader_email': 'uploader@email'
     }, metadata._to_dict())
+
+  def test_trusted_uploader_post(self):
+    """Test trusted uploader post."""
+    data_types.ExternalUserPermission(
```
Comment (oliverchang) [inline src/clusterfuzz/_internal/tests/appengine/handlers/upload_testcase_test.py:493]: You may need to add the email to https://github.com/google/clusterfuzz/blob/7f94ec184870762c04046b23c24ff1a0424133b4/src/clusterfuzz/_internal/datastore/data_types.py#L787 as well. 

Look in other tests for examples of setting up fake Config objects.

File: src/clusterfuzz/_internal/tests/appengine/handlers/upload_testcase_test.py [hunk 4]
```diff
@@ -487,3 +487,29 @@ def test_external_duplicate(self):
...
+      response = client.post(
+          '/',
+          data={
+              'job': 'libfuzzer_proj_external',
+              'target': 'target',
+              'stacktrace': stacktrace,
+              'revision': 1337,
+              'file': (io.BytesIO(b'contents'), 'file'),
+          })
+
```
[the remark this answers is not in the document]
Comment (oliverchang) [reply src/clusterfuzz/_internal/tests/appengine/handlers/upload_testcase_test.py:515]: I see you added a check for he testcase entity, but we also need to get the FuzzTarget entity (https://github.com/google/clusterfuzz/blob/7f94ec184870762c04046b23c24ff1a0424133b4/src/clusterfuzz/_internal/datastore/data_types.py#L1187) and make sure it's what we expect.

File: src/clusterfuzz/_internal/tests/appengine/handlers/upload_testcase_test.py [hunk 5]
```diff
@@ -586,3 +586,34 @@ def test_trusted_uploader_post(self):
...
+        'timeout': 0,
+        'timestamp': datetime.datetime(2021, 1, 1, 0, 0),
+        'uploader_email': 'uploader@email'
+    }, metadata._to_dict())
+    self.assertEqual({
+        'binary': 'target',
+        'engine': 'libFuzzer',
+        'project': 'proj'
+    },
+                     testcase.get_fuzz_target()._to_dict())
```
Comment (oliverchang) [inline src/clusterfuzz/_internal/tests/appengine/handlers/upload_testcase_test.py:619]: Can you query for this FuzzTarget entity directly from the datastore?

The `get_fuzz_target` method for Testcase can construct a fake one on demand in cases where it does not exist: https://github.com/google/clusterfuzz/blob/7f94ec184870762c04046b23c24ff1a0424133b4/src/clusterfuzz/_internal/datastore/data_types.py#L723

File: src/clusterfuzz/_internal/tests/appengine/handlers/upload_testcase_test.py [hunk 6]
```diff
@@ -586,3 +586,34 @@ def test_trusted_uploader_post(self):
...
+        'quiet_flag': False,
+        'retries': None,
+        'security_flag': False,
+        'status': 'Pending',
+        'testcase_id': 5,
+        'timeout': 0,
+        'timestamp': datetime.datetime(2021, 1, 1, 0, 0),
+        'uploader_email': 'uploader@email'
+    }, metadata._to_dict())
+    self.assertEqual({
```
Comment (Alan32Liu) [inline src/clusterfuzz/_internal/tests/appengine/handlers/upload_testcase_test.py:614]: Would using `assertDictEqual` (like the one above) here [make debugging simpler](https://stackoverflow.com/questions/34414326/why-is-assertdictequal-needed-if-dicts-can-be-compared-by)?

**Review by oliverchang — APPROVED** [review]

**Review by Alan32Liu — APPROVED** [review]
LGTM!

**Pull request merged**